src/AutoRest.CSharp/LowLevel/Output/DpgOutputLibraryBuilder.cs

```diff
@@ -385,7 +411,22 @@ private class ClientInfo
             public string Description { get; }
             public INamedTypeSymbol? ExistingType { get; }
             public IReadOnlyList<InputOperation> Operations { get; }
-            public IReadOnlyList<InputParameter> ClientParameters { get; }
+
+            private IReadOnlyList<InputParameter> _clientParameters;
+            public IReadOnlyList<InputParameter> ClientParameters
+            {
+                get
+                {
+                    if (_clientParameters != null && _clientParameters.Count > 0)
+                    {
+                        return _clientParameters;
+                    } else
+                    {
+                        var endpointParameter = this.Children.SelectMany(c => c.ClientParameters).FirstOrDefault(p => p.IsEndpoint);
+                        return endpointParameter != null ? new[] { endpointParameter } : Array.Empty<InputParameter>();
+                    }
+                }
+            }
```

Comment:
To ensure that we only calculate this only once, we should do this in this way (please note that `_clientParameters` is never null, otherwise the compiler will complain. And this is the reason we cannot use it as the backing field of this property):
```suggestion
            private IReadOnlyList<InputParameter> _clientParameters;
            private IReadOnlyList<InputParameter>? _actualClientParameters; // please feel free to name this whatever you think is fine
            public IReadOnlyList<InputParameter> ClientParameters => _actualClientParameters ??= EnsureClientParameters();
            
            private IReadOnlyList<InputParameter> EnsureClientParameters()
            {
                    if (_clientParameters.Count == 0)
                    {
                        var endpointParameter = this.Children.SelectMany(c => c.ClientParameters).FirstOrDefault(p => p.IsEndpoint);
                        return endpointParameter != null ? new[] { endpointParameter } : Array.Empty<InputParameter>();
                    }
                    
                    return _clientParameters;
            }
```

Reply:
updated